Here, scripts used for exploration of RiboGrove data are located.

# Content

1. `count_bases_whole_genome.py` -- the script counts all IUPAC bases in genome sequences in .gbk files

2. `seq_logos/logo_making.ipynb` -- a Jupyter notebook used for sequence logo creation.

3. `alt_antiSD_tests.R` -- a script used for testing is a "mixed" 16S rRNA gene set is uniformly distributed accross the bacterial phyla.

4. `plots/draw_GC_content_plot.R` and `plots/draw_skew_plot.R` -- scripts used to create plots of GC contents and AT and GC skews, respectively.
